File: DSA/VenikaPanwar_590013755/Day_50_Question_2.java
import java.util.*;

public class CountAnagramOccurrences {
    public static int countAnagrams(String text, String pattern) {
        int[] patCount = new int[26];
        int[] winCount = new int[26];
        int count = 0;
        int k = pattern.length();

        for (char c : pattern.toCharArray())
            patCount[c - 'a']++;

        for (int i = 0; i < text.length(); i++) {
            winCount[text.charAt(i) - 'a']++;

            if (i >= k) {
                winCount[text.charAt(i - k) - 'a']--; // remove oldest char
            }

            if (Arrays.equals(patCount, winCount)) {
                count++;
            }
        }

        return count;
    }

    public static void main(String[] args) {
        String text = "forxxorfxdofr";
        String pattern = "for";
        System.out.println("Anagram occurrences: " + countAnagrams(text, pattern));
    }
}

File: DSA/pranavdixit_590014418/Day_50_Question_2.c
#include <stdio.h>
#include <string.h>

#define CHAR 256

int countAnagrams(char* text, char* pattern) {
    int count = 0;
    int lenT = strlen(text), lenP = strlen(pattern);

    int countP[CHAR] = {0}, countW[CHAR] = {0};

    for (int i = 0; i < lenP; i++) {
        countP[pattern[i]]++;
        countW[text[i]]++;
    }

    for (int i = lenP; i < lenT; i++) {
        int match = 1;
        for (int j = 0; j < CHAR; j++) {
            if (countP[j] != countW[j]) {
                match = 0;
                break;
            }
        }
        if (match) count++;

        countW[text[i]]++;
        countW[text[i - lenP]]--;
    }

    int match = 1;
    for (int j = 0; j < CHAR; j++) {
        if (countP[j] != countW[j]) {
            match = 0;
            break;
        }
    }
    if (match) count++;

    return count;
}

int main() {
    char text[] = "forxxorfxdofr";
    char pattern[] = "for";

    int result = countAnagrams(text, pattern);
    printf("Anagram occurrences: %d\n", result);
    return 0;
}
